<a href="https://colab.research.google.com/github/angelostefanini/data_analytics/blob/main/penguins_regressione_multivariata.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [26]:
# Import packages
import pandas as pd
import seaborn as sns

# Load dataset
penguins = sns.load_dataset("penguins")

# Examine first 5 rows of dataset
penguins.head()

species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    Male  
1       3800.0  Female  
2       3250.0  Female  
3          NaN     NaN  
4       3450.0  Female

In [27]:
print(penguins.columns)

Index(['species', 'island', 'bill_length_mm', 'bill_depth_mm',
       'flipper_length_mm', 'body_mass_g', 'sex'],
      dtype='object')


In [28]:
# Drop rows with missing values
penguins.dropna(inplace=True)

# Reset index
penguins.reset_index(inplace=True, drop=True)

In [31]:
# Subset del DataFrame selezionando solo le colonne di interesse
penguins = penguins[["body_mass_g", "bill_length_mm", "sex", "species"]]

# Rinomina le colonne per una migliore chiarezza
penguins = penguins.rename(columns={"sex": "gender"})


In [32]:
# Drop rows with missing values
penguins.dropna(inplace=True)

# Reset index
penguins.reset_index(inplace=True, drop=True)

In [33]:
print(penguins)

     body_mass_g  bill_length_mm  gender species
0         3750.0            39.1    Male  Adelie
1         3800.0            39.5  Female  Adelie
2         3250.0            40.3  Female  Adelie
3         3450.0            36.7  Female  Adelie
4         3650.0            39.3    Male  Adelie
..           ...             ...     ...     ...
328       4925.0            47.2  Female  Gentoo
329       4850.0            46.8  Female  Gentoo
330       5750.0            50.4    Male  Gentoo
331       5200.0            45.2  Female  Gentoo
332       5400.0            49.9    Male  Gentoo

[333 rows x 4 columns]


In [ ]:
# Subset X and y variables
penguins_X = penguins[["bill_length_mm", "gender", "species"]]
penguins_y = penguins[["body_mass_g"]]

In [ ]:
# Import train-test-split function from sci-kit learn
from sklearn.model_selection import train_test_split
# Create training data sets and holdout (testing) data sets
X_train, X_test, y_train, y_test = train_test_split(penguins_X, penguins_y,
                                                    test_size = 0.3, random_state = 42)

In [ ]:
# Write out OLS formula as a string
ols_formula = "body_mass_g ~ bill_length_mm + C(gender) + C(species)"

In [ ]:
from statsmodels.formula.api import ols

In [ ]:
# Create OLS dataframe
ols_data = pd.concat([X_train, y_train], axis = 1)

# Create OLS object and fit the model
OLS = ols(formula = ols_formula, data = ols_data)
model = OLS.fit()

In [ ]:
# Get model results
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            body_mass_g   R-squared:                       0.850
Model:                            OLS   Adj. R-squared:                  0.847
Method:                 Least Squares   F-statistic:                     322.6
Date:                Fri, 19 Jan 2024   Prob (F-statistic):           1.31e-92
Time:                        18:28:12   Log-Likelihood:                -1671.7
No. Observations:                 233   AIC:                             3353.
Df Residuals:                     228   BIC:                             3371.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
===========================================================================================
                              coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------------
Intercept                2032.2111    354.087      5.739      0.000    1334.510    2729.913
C(gender)[T.Male]         528.9508     55.105      9.599      0.000     420.371     637.531
C(species)[T.Chinstrap]  -285.3865    106.339     -2.684      0.008    -494.920     -75.853
C(species)[T.Gentoo]     1081.6246     94.953     11.391      0.000     894.526    1268.723
bill_length_mm             35.5505      9.493      3.745      0.000      16.845      54.256
==============================================================================
Omnibus:                        0.339   Durbin-Watson:                   1.948
Prob(Omnibus):                  0.844   Jarque-Bera (JB):                0.436
Skew:                           0.084   Prob(JB):                        0.804
Kurtosis:                       2.871   Cond. No.                         798.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Coefficienti (Coef):
I coefficienti sono valori che rappresentano la variazione attesa nella variabile dipendente per una variazione unitaria nella variabile indipendente corrispondente, mantenendo costanti tutte le altre variabili. Nel tuo modello:

Intercept: Questo è il valore previsto della variabile dipendente quando tutte le variabili indipendenti sono zero. Nel tuo caso, l'intercetta è 2032.2111.

C(gender)[T.Male]: Questo coefficiente è associato alla variabile categorica "gender", con la categoria di riferimento implicita come femmina. Il valore 528.9508 indica la differenza attesa nella variabile dipendente tra i maschi e le femmine, mantenendo costanti le altre variabili.

C(species)[T.Chinstrap]: Questo coefficiente è associato alla variabile categorica "species" con la categoria di riferimento implicita come "Adelie". Il valore -285.3865 indica la differenza attesa nella variabile dipendente tra la specie "Chinstrap" e "Adelie", mantenendo costanti le altre variabili.

C(species)[T.Gentoo]: Questo coefficiente è associato alla variabile categorica "species" con la categoria di riferimento implicita come "Adelie". Il valore 1081.6246 indica la differenza attesa nella variabile dipendente tra la specie "Gentoo" e "Adelie", mantenendo costanti le altre variabili.

bill_length_mm: Questo coefficiente è associato alla variabile continua "bill_length_mm". Il valore 35.5505 indica la variazione attesa nella variabile dipendente per una variazione unitaria nella lunghezza del becco, mantenendo costanti le altre variabili.

P>|t| (Valori p dei t-test):
I valori p dei t-test forniscono informazioni sulla significatività statistica di ciascun coefficiente. I valori p indicano la probabilità di ottenere un t-value almeno tanto estremo quanto quello osservato, supponendo che il coefficiente associato sia in realtà zero (nessun effetto). Nel tuo caso:

P>|t| per tutti i coefficienti: Tutti i valori p sono molto vicini a zero (0.000), il che indica che tutti i coefficienti sono considerati statisticamente significativi. In altre parole, c'è evidenza sufficiente per respingere l'ipotesi nulla che il coefficiente sia zero.
La significatività statistica dei coefficienti è importante per capire quali variabili contribuiscono in modo significativo alla predizione della variabile dipendente nel contesto del tuo modello di regressione lineare.
